plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,545 @@
+		g_debug ("out_max_window_size: %d\n", (int)out_max_window_size);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+
+	total_sending_numbers = (g_bytes_get_size (ctx->blob)) / (out_max_transfer_size);
+	if ((g_bytes_get_size (ctx->blob)) % (out_max_transfer_size) != 0)
+		total_sending_numbers ++;
+
+	ctx->chunks = fu_chunk_array_new_from_bytes (ctx->blob,
+						0x00,	/* start addr */
+						0x00,	/* page_sz */
+						out_max_transfer_size);
+	chk = g_ptr_array_index (ctx->chunks, 0);
+	request = (mbim_message_qdu_file_write_set_new (fu_chunk_get_data_sz (chk), (const guint8 *)fu_chunk_get_data (chk), &error));
+	mbim_device_command (ctx->mbim_device,
+			     request,
+			     10,
+			     ctx->cancellable,
+			     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_write_ready,
+			     ctx);
+	mbim_message_unref (request);
+}
+
+static void
+fu_mbim_qdu_updater_session_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	MbimMessage *response = NULL;
+	MbimQduSessionType     out_current_session_type;
+	MbimQduSessionStatus   out_current_session_status;
+	MbimQduSessionType     out_last_session_type;
+	MbimQduSessionResult   out_last_session_result;
+	guint32                out_last_session_error_offset;
+	guint32                out_last_session_error_size;
+	g_autoptr(GError)      error = NULL;
+	MbimMessage *request = NULL;
+	WriteContext *ctx = user_data;
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_printerr ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_update_session_response_parse (
+							     response,
+							     &out_current_session_type,
+							     &out_current_session_status,
+							     &out_last_session_type,
+							     &out_last_session_result,
+							     &out_last_session_error_offset,
+							     &out_last_session_error_size,
+							     &error)) {
+		g_printerr ("error: couldn't parse response message: %s\n", error->message);
+		g_printerr ("out_current_session_type: %d\n", (int)out_current_session_type);
+		g_printerr ("out_current_session_status: %d\n", (int)out_current_session_status);
+		g_printerr ("out_last_session_type: %d\n", (int)out_last_session_type);
+		g_printerr ("out_last_session_result: %d\n", (int)out_last_session_result);
+		g_printerr ("out_last_session_error_offset: %d\n", (int)out_last_session_error_offset);
+		g_printerr ("out_last_session_error_size: %d\n", (int)out_last_session_error_size);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+	g_debug ("[%s] Successfully request modem to update session", mbim_device_get_path_display (device));
+
+	request = (mbim_message_qdu_file_open_set_new (0, g_bytes_get_size (ctx->blob), &error));
```

Comment:
Why the extra () around the method call?
Also just pass NULL at the GError, the set_new() method should never fail.